src/rgw/rgw_op.cc

```diff
@@ -6139,6 +6144,81 @@ void RGWCompleteMultipart::execute(optional_yield y)
   }
 }
 
+bool RGWCompleteMultipart::check_previously_completed(const DoutPrefixProvider* dpp, const RGWMPObj& mpobj, const RGWMultiCompleteUpload* parts, rgw::sal::Store* store)
+{
+  // TODO: replace with mpstat() in zipper
+  auto rados = dynamic_cast<rgw::sal::RadosStore*>(store);
+  if (!rados) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: rados not avaliable" << dendl;
+    return false;
+  }
+  if (parts->parts.empty()) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: parts list is empty" << dendl;
+    return false;
+  }
+  // check that a part with the same upload-id exists
+  std::string oid = mpobj.get_part(parts->parts.begin()->first);
+  rgw_obj chk_obj;
+  chk_obj.init_ns(s->bucket->get_key(), oid, mp_ns);
+  rgw_obj_index_key key;
+  chk_obj.key.get_index_key(&key);
+  rgw_pool pool;
+
+  bool bret = rados->getRados()->get_obj_data_pool(s->bucket->get_placement_rule(), chk_obj, &pool);
+  if (!bret) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: get_obj_data_pool() returned bret=" << bret << dendl;
+    return false;
+  }
+  int ret;
+  librados::IoCtx data_ioctx;
+  ret = rgw_init_ioctx(rados->getRados()->get_rados_handle(), pool, data_ioctx);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: rgw_init_ioctx() returned ret=" << ret << dendl;
+    return false;
+  }
+  std::string chk_oid = s->bucket->get_bucket_id() + "_" + key.name;
+  ldpp_dout(dpp, 5) << __func__ << "() verifying existance of previous multipart part: " << chk_oid << dendl;
+  ret = data_ioctx.stat(chk_oid, nullptr, nullptr);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: ioctx.stat(" << chk_oid << ") returned ret = " << ret << dendl;
+    return false;
+  }
+  // re-calculate the etag from the parts and compare to the existing object
+  s->object->set_bucket(s->bucket.get());
+  ret = s->object->get_obj_attrs(s->obj_ctx, s->yield, this);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: get_obj_attrs() returned ret=" << ret << dendl;
+    return false;
+  }
+  rgw::sal::Attrs sattrs = s->object->get_attrs();
+  string oetag = sattrs[RGW_ATTR_ETAG].to_str();
+
+  MD5 hash;
+  for (auto partiter = parts->parts.begin() ; partiter != parts->parts.end() ; ++partiter) {
```

Comment:
maybe consider the ranged-for syntax here? it's a bit shorter, and lets you give better names than `first` and `second`. up to you
```c++
for (const auto& [index, part] : parts->parts) {
```

Reply:
thank you, done